src/coreclr/vm/jitinterface.cpp

```diff
@@ -3624,10 +3624,12 @@ uint32_t CEEInfo::getClassAttribsInternal (CORINFO_CLASS_HANDLE clsHnd)
             if (pMT->IsByRefLike())
                 ret |= CORINFO_FLG_BYREF_LIKE;
 
-            if ((pClass->IsNotTightlyPacked() && (!pClass->IsManagedSequential() || pClass->HasExplicitSize())) ||
+            if ((pClass->IsNotTightlyPacked() && (pClass->HasExplicitSize())) ||
                 pMT == g_TypedReferenceMT ||
                 VMClsHnd.IsNativeValueType())
             {
+                // All kinds of structs with Size=.., Pack=... will be marked as "Explicit Size"
+                // and end up as "custom layout". We consider "Auto-layout" as non-custom one
```

Comment:
> We consider "Auto-layout" as non-custom one

Is this an okay assumption?

Reply:
Actually, it might be not a correct assumption 🤔 let me just mirror crossgen logic

Reply:
This condition is very fragile. The meaning of `CORINFO_FLG_CUSTOMLAYOUT` is very poorly defined. I do not think you want to be touching it to get the regression fixed quickly.

It sounds like that some flags on `MethodTable` changed and that introduced the regression. What are the flags on `MethodTable` that changed?

Reply:
@jkotas in JIT we have a logic:
```
    if (StructHasCustomLayout(typeFlags) && ((typeFlags & CORINFO_FLG_CONTAINS_GC_PTR) == 0))
    {
        structPromotionInfo.customLayout = true;
    }
```
our assumption that with the previous impl it was not reported as `CORINFO_FLG_CONTAINS_GC_PTR` with e.g. ByReference<byte>

Reply:
or, actually, the other way. I am still digging.

Reply:
> our assumption that with the previous impl it was not reported as CORINFO_FLG_CONTAINS_GC_PTR with e.g. ByReference<byte>

That seems like we would have had much larger issues. The JIT did special case `ByReference<byte>` in a few places and I removed them all. Perhaps something was missed in indicating the type was reducible to a byref?

Reply:
@jkotas @AaronRobinsonMSFT 

```c
pMT->ContainsPointers();  // was: 0, now: 0
pClass->IsManagedSequential(); // was: 1, now: 0
pClass->HasExplicitSize(); // was: 0, now: 0
pClass->IsAutoLayoutOrHasAutoLayoutField(); // was: 0, now: 0
pMT->dwFlags // was: 266256, now: 266256
```
Seems like the only thing changed is `IsManagedSequential` to `0` that with current logic leads to `CORINFO_FLG_CUSTOMLAYOUT ` sent to JIT.
From my understanding, the primary usage for this flag is to tell jit that e.g. holes in current struct are likely "valueable" so it better not promote it.

As I noted earlier, the check is different in crossgen (looks better for jit): https://github.com/dotnet/runtime/blob/2170ea9d375baac555dc348d85cc6b8fed10aef8/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs#L2097-L2099

Reply:
> Seems like the only thing changed is IsManagedSequential to 0 that with current logic leads to

So the simple fix should be to change this back. I think that the following change in `TypeHasGCPointers` is going to do that:

```
< if (CorTypeInfo::IsPrimitiveType(corElemType) || corElemType == ELEMENT_TYPE_PTR || corElemType == ELEMENT_TYPE_FNPTR)
> if (CorTypeInfo::IsPrimitiveType(corElemType) || corElemType == ELEMENT_TYPE_PTR || corElemType == ELEMENT_TYPE_FNPTR || corElemType == ELEMENT_TYPE_BYREF)
```

The custom layout flag is a mess and it would be nice to revisit it, but it should be done in a separate PR.

Reply:
We no longer classify ``System.Span`1`` as managed sequential because `TypeHasGCPointers(ELEMENT_TYPE_BYREF, ...)` returns true, so we set `fDisqualifyFromManagedSequential` in classlayoutinfo.cpp. Before the change ``TypeHasGCPointers(ELEMENT_TYPE_VALUETYPE, <type handle for System.ByReference`1>)`` was returning false.

Reply:
Notice that the logic in https://github.com/dotnet/runtime/blob/27195f670937c7e21ab68a806396f9d17c57231a/src/coreclr/vm/classlayoutinfo.cpp#L304-L312 is inconsistent right now. It only returns true for GC pointers (that do not include byrefs) for nested types, but it returns true for both GC pointers and byrefs when they are contained directly.

The change that I have proposed above is going to fix this inconsistency.

Reply:
> The custom layout flag is a mess and it would be nice to revisit it, but it should be done in a separate PR.

I have opened https://github.com/dotnet/runtime/issues/71711 to track this.